### ISSUE TITLE
RFC: API Changes

### ISSUE THREAD

**Issue by jamesarosen** [minutebase/ember-can#27]
A while back, @rlivsey asked me if I had any ideas on how to improve the API for this library. I didn't at the time because I didn't have enough concrete cases. I've since found a few things that are harder than they should be.

### Prevents Cohesion

In `route:posts`, we check whether the user has permission to manage posts in general:
```js
if (!this.can("manage posts")) { this.replaceWith('index'); }
```

In `route:post/edit`, we check whether the user has permission to manage that specific post:
```js
if (!this.can("manage post", post)) { this.replaceWith('index'); }
```

But we have to define those two abilities in `app/abilities/posts.js` and `app/abilities/post.js` even though the logic is closely related.

### Prevents Cohesion, Part 2

We have about 20 permissions that related to about 12 different models. All of the definitions are very short and they all look very similar. Having `app/billing/ability.js` and `app/account/ability.js` just to get names seems silly. It would be easier if we could just do

```js
// app/abilities/all.js
export default Ability.extend({
  canManageBilling: ...,
  canManageAccount: ...,
});
```

### Makes Searching Hard

Permissions are defined as `FooAbility.prototype.canTwizzleBar`, but used as `can("twizzle bar in foo")`. This makes it hard to find the definition of the thing you're looking at. If I see a template that looks like
```hbs
{{#if (can "twizzle bar in foo")}}
```
I want to copy and paste that string and find the definition. That would be easier if it were
```hbs
{{#if (can "twizzleBarInFoo")}}
```
*or* if the definition were
```js
Ability.extend({
  "can twizzle bar in foo": function() { return true; }
});
```

### Suggestion

I don't have a single concrete suggestion. I understand the value of having a `FooAbility` with just a `canEdit`, especially for projects that have many or complex permissions schemes.

My inclination would be to have apps define a single `app/abilities.js`. If larger apps want to break that logic up, they can.

An alternative would be to have apps define any number of `app/abilities/*.js`, but each is a `Mixin`. They all get combined into one object.

My two alternatives both require moving the full name of the ability into the property key instead of composing the key and the class name.

**Comment by rlivsey** [comment]
Had a discussion with @jamesarosen on Slack:

```
jamesarosen [8:55 PM] 
I really wish I had some big inspiration that let me do what I want without compromising the organization for the larger projects

jamesarosen [8:56 PM]
it helps that there's almost no state to these objects, though

rlivsey [8:58 PM] 
yeah, I wonder if (like you suggest) there’s a way of starting off with a base/all/default ability which you can then split up into more specific ones later with it still falling back to the default if it doesn’t find it in the specific location

jamesarosen [8:59 PM] 
that's a neat idea

rlivsey [9:00 PM] 
might be over-complex, but could also have a resolver which is able to be overridden easily. The default resolver would map to individual files but could easily be overridden to do something simpler / more complex based on the app

jamesarosen [9:02 PM] 
Or even just make the `resolveName(s) => { klass, property }` function part of the public API so it can be overridden in an initializer.

rlivsey [9:03 PM] 
yep that could work well

rlivsey [9:05 PM]
would be an easy change & a nice way of being able to experiment with the API in a backwards compatible way, then we’ll probably discover a much better default API
```

**Comment by rlivsey** [comment]
https://github.com/minutebase/ember-can#custom-ability-lookup

**Comment by rlivsey** [comment]
Released as 0.7.0.

I've converted my app to use the above example and it's really nice that autocompletion works & no more mental parsing of ability strings is needed.

Some experimentation needed, but I think it'll be worth making something like that the default for v1.0 with a deprecation warning in v0.8.x for the current behaviour. Ideally with an ember-watson plugin to auto-transform existing apps.
